api/v1alpha1/lvmcluster_types.go

```diff
@@ -70,6 +77,17 @@ type DeviceClass struct {
 	// +kubebuilder:validation:Pattern="^[a-z0-9]([-a-z0-9]*[a-z0-9])?$"
 	Name string `json:"name,omitempty"`
 
+	// DeviceDiscoveryPolicy is a flag to indicate whether the device should be discovered
+	// at install time or at runtime (static or dynamic configuration of devices)
+	// If set to DeviceDiscoveryPolicyInstallStatic, the devices will not be added to the VG if they are not present at setup time.
+	// If set to DeviceDiscoveryPolicyRuntimeDynamic, the devices will be added to the VG if they are present at runtime.
+	// By default, the value is set to RuntimeDynamic.
+	// This field cannot be updated once the LVMCluster is created.
+	// +kubebuilder:validation:Enum=InstallStatic;RuntimeDynamic
```

Comment:
I thought we had something in the OpenShift Conventions about Enums but as it turns out, it was some of my prior art that triggered the thought :smile: 

Drawing on a distant memory, the OpenShift API has a concept of an intentional "empty" or null choice for Enums so a cluster admin can choose to set a value, and then unset it to allow whatever the default is take effect. The main use case as I recall is upgrades where a cluster admin may want to take advantage of a new way of configuring a field or no longer wants it configured. Is that something that'd be valuable here?

Here are some references:
* https://github.com/openshift/enhancements/blob/master/enhancements/dns/forward-dns-over-tls.md
* https://github.com/openshift/api/pull/1110/files#diff-46e86efb5c123e387737010ac03828fda0c5edfa402c432480485d928827236cR135

Reply:
TBH thats the first time I've heard about this so let me try to understand what youre trying to achieve:

Assuming the LVMCluster is configured
1. The fields would be set to some value
2. The administrator would want to use a new default, and set it to ""
3. The administrator would rely on OpenShift to figure out the new default.

Is that correct? If so im not sure if its applicable, because the Field is immutable after creation of the LVMCluster, meaning changing the value is impossible (at least as its implemented right now). Otherwise we would have to test switching from one setup to the other.

Maybe im missing things though. Either way we can also implement a defaulting logic like you suggested I think. Defaulting to @suleymanakbas91 opinion to break the tie here.

Reply:
Ah one more thing, if in a future release the default should change, we would change the CRD annotation accordingly, I don't think there is a mandatory need for a zero-value here.

Reply:
+1 to having @suleymanakbas91 tie break although now that I think about it, I guess a better question from me would be, is there a use case for a mutable value?

After relooking at my reference I realized the the field was optional, but that's not the case here so it might not even apply. The only thing I could think of which is probably a weak justification is "useability" in that the field could be hidden if it's not set, giving users less configuration to key in on. Perhaps a GitOps impacting use case? What do you think?

P.S. I'll resolve the other comments around this since whatever gets decided will have to affect the references anyway.

Reply:
On the GitOps part: When applying via e.g. ArgoCD, it will use either SSA or Strategic Merge Patch, in which case fields that are not set in the Argo Specification but are defaulted in through the operator, are ignored by Argo. Flux should behave the same / similar.

Reply:
For the mutable value: The use case would be that a user deploys with an initially static configuration, but then has the need to expand the storage later on with some devices he does not know the serials from. I think this is too much of an edge case to consider for the implementation now, so thats why I left it as immutable

Reply:
Restating to make sure I understand; we're saying that it's an edge use case for a cluster admin to go from static configuration with a known set of disks to dynamic with an unknown set of disks?

Reply:
Yeah that pretty much sums it up. I think for people who want to try out LVMS they usually start with dynamic configuration, only to later find out that they want to have a static device discovery

Reply:
Ah okay, that makes sense. I'm comfortable resolving this thread with the understanding that I'm unsure if practically there's anything we need to future guard against or plan for -- the rationale for the implementation is clear.